sdformat_urdf parser demo

quarkytale opened the pull request:
# 🎉 New demo

## Summary
Demonstrating SDFormat XML as a robot description format instead of URDF XML using [sdformat_urdf](https://github.com/ros/sdformat_urdf) plugin. The parser takes the SDF file published on `/robot_description` topic by [robot_state_publisher](https://github.com/ros/robot_state_publisher) and converts it to URDF C++ DOM structures. Joint states and TFs are bridged using `ros_ign_bridge` enabling RViz to visualize the RobotModel and odometry simulated by a Gazebo world.
Also demonstrates a use-case of #259.

### Changes made
- Parser compatible SDF model
- Gazebo world with required plugins (JointStatePublisher, PosePublisher, OdometryPublisher)
- Launch file bringing up launch arguments, Gazebo, ros_ign_bridge, robot_state_publisher and RViz
- RViz config
- Installation framework for the new model and world

### Snapshots
![vehicle](https://user-images.githubusercontent.com/24695820/175471702-efabdd05-51c8-46c0-afe3-e77ce0e81689.png)

https://user-images.githubusercontent.com/24695820/176565497-80414260-f958-44cc-8323-673077fad3d8.mp4



## Test it
In a colcon workspace, install `ros_gz` and `sdformat_urdf` from source. Make sure you have the latest Gazebo fortress release.
Run the demo launch file with the rviz launch argument set.
```
ros2 launch ros_ign_gazebo_demos sdf_parser.launch.py rviz:=True
```
Start the simulation in Gazebo and wait a second for TFs to be published.
Send commands for the vehicle to move in circles, matching its trajectory in Gazebo and RViz.
```
ign topic -t "/model/vehicle/cmd_vel" -m ignition.msgs.Twist -p "linear: {x: 1.0}, angular: {z: -0.1}"
ros2 topic pub /model/vehicle/cmd_vel geometry_msgs/msg/Twist "{linear: {x: 5.0, y: 0.0, z: 0.0}, angular: {x: 0.0, y: 0.0, z: -0.1}}
```

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [x] Added example and/or tutorial
- [x] Updated documentation (as needed)
- [x] Updated migration guide (as needed)
- [x] Consider updating Python bindings (if the library has them)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

methylDragon committed:
Move packages and files to gz

Signed-off-by: methylDragon <methylDragon@gmail.com>

quarkytale committed:
parser compatible model and launch framework

Signed-off-by: Dharini Dutia <dharini@openrobotics.org>

quarkytale committed:
added ground plane, common gz plugins, demo commands and cleaned install paths

Signed-off-by: Dharini Dutia <dharini@openrobotics.org>

quarkytale committed:
unique collision names and cleared flake

Signed-off-by: Dharini Dutia <dharini@openrobotics.org>

mjcarroll commented:
I'm pretty sure CI is consistently broken on the `ros2` branch.

Commit:
Merge branch 'ros2' into quarkytale/parser_demo

Signed-off-by: Michael Carroll <michael@openrobotics.org>

chapulina commented on ros_ign_gazebo_demos/models/vehicle/model.config:
```diff
@@ -0,0 +1,13 @@
+  </author>
+
+  <description>
+  Parser compatible differential drive vehicle model.
```
It would be good to explain what parser this is referring to

chapulina commented on ros_ign_gazebo_demos/models/vehicle/model.config:
```diff
@@ -0,0 +1,13 @@
+  <sdf version="1.8">model.sdf</sdf>
+
+  <author>
+  </author>
```
It's always good to add an author

chapulina reviewed (changes requested):
This PR should help with `sdformat_urdf` + Garden. We need to build that from source with Garden enabled on this project's CI.

* https://github.com/ros/sdformat_urdf/pull/17

quarkytale committed:
updating model config

Signed-off-by: Dharini Dutia <dharini@openrobotics.org>

quarkytale committed:
building parser from source

Signed-off-by: Dharini Dutia <dharini@openrobotics.org>

quarkytale committed:
fix flake and update deb dependency for garden

Signed-off-by: Dharini Dutia <dharini@openrobotics.org>

chapulina commented on ros_ign_gazebo_demos/CMakeLists.txt:
```diff
@@ -27,4 +27,18 @@ install(
 
+install(
+  DIRECTORY
+    ${CMAKE_CURRENT_SOURCE_DIR}/models/vehicle
```
How about installing the entire `models` folder? This way we don't need to update this when we add new models.

chapulina commented on ros_ign_gazebo_demos/package.xml:
```diff
@@ -24,6 +24,7 @@
   <exec_depend>rqt_plot</exec_depend>
   <exec_depend>rqt_topic</exec_depend>
   <exec_depend>rviz2</exec_depend>
+  <exec_depend condition="$IGNITION_VERSION != garden">>sdformat_urdf</exec_depend>
```
Should this be

```suggestion
  <exec_depend condition="$IGNITION_VERSION == garden">sdformat_urdf</exec_depend>
```

quarkytale replied:
Removed the extra > but the condition is correct, since we don't want to install from binaries if its garden

chapulina replied:
Oh I see what you're saying. The `package.xml` file isn't used just by `rosdep` to install binary dependencies though. It's also used by `colcon` and `bloom`.

Excluding Garden here gives the impression that the demos that use `sdformat_urdf` can't be used with Garden, which is not true. So we should be able to leave this condition out. CI is running `rosdep` with the `-r` flag, so it won't fail when it can't find this dependency.

quarkytale replied:
CI was failing without it though
```
ERROR: the following rosdeps failed to install
  apt: command [apt-get install -y ros-rolling-sdformat-urdf] failed
  apt: Failed to detect successful installation of [ros-rolling-sdformat-urdf]
```

chapulina replied:
It looks like the issue here is a bit higher:

```
gz-tools2 : Breaks: ignition-tools but 1.5.0-1~jammy is to be installed
```

We shouldn't be pulling `ignition-tools` on the Garden job though, only `gz-tools`. Not sure yet what's pulling that.

chapulina replied:
`libsdformat12-dev` is pulling that, see this comment https://github.com/gazebo-tooling/release-tools/issues/529#issuecomment-1231016550

chapulina replied:
Ok, so instead of trying to suppress the installation of `sdformat_urdf` here, I propose we do it just for CI at `build-and-test.sh`. Something like this (untested):

```diff
diff --git a/.github/workflows/build-and-test.sh b/.github/workflows/build-and-test.sh
index 1ac4206..aa516d1 100755
--- a/.github/workflows/build-and-test.sh
+++ b/.github/workflows/build-and-test.sh
@@ -17,6 +17,8 @@ if [ "$GZ_VERSION" == "garden" ]; then
   wget https://packages.osrfoundation.org/gazebo.key -O - | apt-key add -
 
   GZ_DEPS="libgz-sim7-dev"
+
+  ROSDEP_ARGS="--skip-keys='sdformat-urdf'"
 fi
 
 # Fortress comes through rosdep for Focal and Jammy
@@ -31,7 +33,7 @@ apt-get install -y $GZ_DEPS \
 
 rosdep init
 rosdep update
-rosdep install --from-paths ./ -i -y -r --rosdistro $ROS_DISTRO
+rosdep install --from-paths ./ -i -y -r --rosdistro $ROS_DISTRO $ROSDEP_ARGS
 
 # Build.
 source /opt/ros/$ROS_DISTRO/setup.bash
```

This way colcon users would still get that dependency.

quarkytale replied:
Seems to have worked, CI is green!

chapulina replied:
Awesome! We can remove the condition here then! This way, folks building with colcon will get the dependencies resolved correctly

```
<exec_depend>sdformat_urdf</exec_depend>
```

quarkytale replied:
All set

chapulina commented on ros_ign_gazebo_demos/worlds/vehicle.sdf:
```diff
@@ -0,0 +1,106 @@
+
+  Try sending commands:
+
+    ign topic -t "/model/vehicle/cmd_vel" -m ignition.msgs.Twist -p "linear: {x: 1.0}, angular: {z: -0.1}"
```
How about sending ROS commands here? It makes this a more compelling ROS demo.

chapulina commented on ros_ign_gazebo_demos/worlds/vehicle.sdf:
```diff
@@ -0,0 +1,106 @@
+    <plugin
+      filename="ignition-gazebo-scene-broadcaster-system"
+      name="ignition::gazebo::systems::SceneBroadcaster">
+    </plugin>
```
You can remove these system plugins, since they're the default ones.

chapulina commented on ros_ign_gazebo_demos/worlds/vehicle.sdf:
```diff
@@ -0,0 +1,106 @@
+<sdf version="1.8">
+  <world name="demo">
+    <physics name="1ms" type="ignored">
+      <real_time_factor>0</real_time_factor>
```
For this kind of demo, I think it's more user-friendly to run at real time, i.e. `1.0`. That's already the default, so we can remove the entire `physics` block.

chapulina commented on ros_ign_gazebo_demos/worlds/vehicle.sdf:
```diff
@@ -0,0 +1,106 @@
+        <uri>package://ros_ign_gazebo_demos/models/vehicle</uri>
+      </include>
+
+    <plugin
```
Nit: indentation

chapulina commented on ros_ign_gazebo_demos/worlds/vehicle.sdf:
```diff
@@ -0,0 +1,106 @@
+      <plugin
+        filename="ignition-gazebo-odometry-publisher-system"
+        name="ignition::gazebo::systems::OdometryPublisher">
+        <dimensions>3</dimensions>
```
For a ground vehicle, we don't need 3 dimensions

chapulina reviewed (changes requested):
The demo isn't working for me on RViz, have you tried the latest state of this branch, @quarkytale ? I haven't dug to see what could be wrong.

methylDragon committed:
Support ros_ign migration

Clean up shared libraries, and tick-tock RosGzPointCloud
Tick-tock launch args
Hard-tock ign_ in sources
Migrate ign, ign_, IGN_ for sources, launch, and test files
Migrate IGN_XXX_VER, IGN_T, header guards
Migrate launchfile, launchfile args, and test source references
Migrate ros_ign_XXX and gz_gazebo -> gz_sim
Migrate ros_ign_XXX project names
Migrate Ign, ign-, IGN_DEPS, ign-gazebo
Migrate ignitionrobotics, ignitionrobotics/ros_ign, osrf/ros_ign
Migrate ignition-version, IGNITION_VERSION, Ignition <LIB>, ros_ign_ci

Signed-off-by: methylDragon <methylDragon@gmail.com>

quarkytale committed:
Merge branch 'ros2' into quarkytale/parser_demo

quarkytale committed:
feedback and ign->gz

Signed-off-by: Dharini Dutia <dharini@openrobotics.org>

quarkytale committed:
Merge branch 'ros_gz_migration-mv' of github.com:gazebosim/ros_gz into quarkytale/parser_demo

quarkytale committed:
renaming and flake

Signed-off-by: Dharini Dutia <dharini@openrobotics.org>

quarkytale committed:
added ros commands

Signed-off-by: Dharini Dutia <dharini@openrobotics.org>

quarkytale commented:
`test_launch_test_ros_subscriber.launch.py` is failing for fortress+humble

quarkytale committed:
Merge branch 'ros2' into quarkytale/parser_demo

chapulina committed:
Merge branch 'ros2' into quarkytale/parser_demo

Signed-off-by: Louise Poubel <louise@openrobotics.org>

chapulina committed:
gz-version

Signed-off-by: Louise Poubel <louise@openrobotics.org>

chapulina commented on ros_gz_sim_demos/launch/sdf_parser.launch.py:
```diff
@@ -0,0 +1,109 @@
+        robot_desc = infp.read()
+
+    rviz_launch_arg = DeclareLaunchArgument(
+        'rviz', default_value='false',
```
I think we could default to true here, the demo is more compelling when RViz is opened.

chapulina commented on ros_gz_sim_demos/worlds/vehicle.sdf:
```diff
@@ -0,0 +1,90 @@
+  Try sending commands:
+
+    gz topic -t "/model/vehicle/cmd_vel" -m gz.msgs.Twist -p "linear: {x: 1.0}, angular: {z: -0.1}"
+    ros2 topic pub /model/vehicle/cmd_vel geometry_msgs/msg/Twist "{linear: {x: 5.0, y: 0.0, z: 0.0}, angular: {x: 0.0, y: 0.0, z: -0.1}}
```
Missing a `"`

```suggestion
    ros2 topic pub /model/vehicle/cmd_vel geometry_msgs/msg/Twist "{linear: {x: 5.0, y: 0.0, z: 0.0}, angular: {x: 0.0, y: 0.0, z: -0.1}}"
```

chapulina commented on ros_gz_sim_demos/launch/sdf_parser.launch.py:
```diff
@@ -0,0 +1,109 @@
+    )
+
+    # Bridge to forward tf and joint states to ros2
+    gz_topic = '/model/vehicle'
```
I needed to move the `/` down to line 55 for this to work

```suggestion
    gz_topic = 'model/vehicle'
```

chapulina commented on ros_gz_sim_demos/launch/sdf_parser.launch.py:
```diff
@@ -0,0 +1,109 @@
+    # Bridge to forward tf and joint states to ros2
+    gz_topic = '/model/vehicle'
+    joint_state_gz_topic = '/world/demo/' + gz_topic + '/joint_state'
+    link_pose_gz_topic = gz_topic + '/pose'
```
```suggestion
    link_pose_gz_topic = '/' + gz_topic + '/pose'
```

chapulina reviewed (commented):
Some minor comments with things I needed to do for this to run

quarkytale committed:
feedback and ci trial

Signed-off-by: Dharini Dutia <dharini@openrobotics.org>

quarkytale committed:
removing garden condition

Signed-off-by: Dharini Dutia <dharini@openrobotics.org>

chapulina reviewed (approved):
This is great! 🚀
